### PR TITLE
Feature/ghi 406 requesting signatures

### LINKED ISSUE #406: Requesting signatures for transaction confirmation

### PR DESCRIPTION
**Pull request by sche**
requesting signatures logic

closes #406

### CODE REVIEW

File: Common/Common/OneOperationWaitinScheduler.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
+//
+//  Copyright © 2018 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+
+/// Allows to schedule operations that will be executed not more often than once in "interval" seconds.
+/// It allows to have only one opetion in a queue to execute.
```
Comment (DmitryBespalov) [inline Common/Common/OneOperationWaitinScheduler.swift]: typo: `opetion` -> `operation`

File: Common/Common/OneOperationWaitinScheduler.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+public final class OneOperationWaitinScheduler {
```
Comment (DmitryBespalov) [inline Common/Common/OneOperationWaitinScheduler.swift]: typo: `Waitin` -> `Waiting`?

File: SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift [hunk 3]
```diff
@@ -131,4 +175,87 @@ final class ReviewTransactionViewController: UITableViewController {
...
+    private func updateConfirmationCell(_ tx: TransactionData) {
+        switch tx.status {
+        case .waitingForConfirmation:
+            confirmationCell.transactionConfirmationView.status = .pending
+        case .readyToSubmit:
+            confirmationCell.transactionConfirmationView.status = .confirmed
+        case .rejected:
+            confirmationCell.transactionConfirmationView.status = .rejected
+        case .success:
+            // TODO: unit test this
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift]: please unit test this

File: SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift [hunk 4]
```diff
@@ -131,4 +175,87 @@ final class ReviewTransactionViewController: UITableViewController {
...
+                }
+            }
+        }
+    }
+
+    // MARK: - Submitting transaction
+
+    @objc internal func submit() {
+        guard tx.status == .readyToSubmit else {
+            showTransactionNeedsConfirmationAlert()
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift:237]: can't we disable the 'submit' button until the tx is ready to submit?

Comment (DmitryBespalov) [reply SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift:237]: or is it a desing requirement for this behavior?

File: SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewControllerTests.swift [hunk 5]
```diff
@@ -74,37 +80,114 @@ class ReviewTransactionViewControllerTests: XCTestCase {
...
+    }
+
+    func test_whenSubmittingConfirmedTransaction_thenCallsDelegate() {
+        let (_, vc) = ethDataAndCotroller(.readyToSubmit)
+        vc.submit()
+        XCTAssertTrue(delegate.requestedToSubmit)
+    }
+
+    // TODO: test_whenSubmittingConfirmedTransactonAndNotAllowed_thenDoesNothing
+    // TODO: test_whenSubmittingConfirmedTransactonAndAllowed_thenCallesDelegateOnSuccess
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewControllerTests.swift]: is this going to be implemented in the next PR?

File: TestTools/XCTest+Alert.swift [hunk 6]
```diff
@@ -6,7 +6,10 @@ import Foundation
 import XCTest
 import UIKit
 
-public func XCTAssertAlertShown(message expectedMessage: String? = nil, file: StaticString = #file, line: UInt = #line) {
+public func XCTAssertAlertShown(message expectedMessage: String? = nil,
+                                actionsCount: Int = 1,
```
Comment (DmitryBespalov) [inline TestTools/XCTest+Alert.swift]: please rename `actionsCount` to `actionCount`

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]

File: Common/Common/OneOperationWaitinScheduler.swift (hunk 1 above)
Comment (sche) [reply Common/Common/OneOperationWaitinScheduler.swift]: done

File: Common/Common/OneOperationWaitinScheduler.swift (hunk 2 above)
Comment (sche) [reply Common/Common/OneOperationWaitinScheduler.swift]: done

File: SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift (hunk 4 above)
Comment (sche) [reply SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewController.swift:237]: Now we designed it like that. Let's see what will change today...

File: SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewControllerTests.swift (hunk 5 above)
Comment (sche) [reply SafeUIKit/SafeAppUI/ReusableDetachedControllers/ReviewTransaction/ReviewTransactionViewControllerTests.swift]: I will implement in this PR

File: TestTools/XCTest+Alert.swift (hunk 6 above)
Comment (sche) [reply TestTools/XCTest+Alert.swift]: done

**Review by DmitryBespalov — APPROVED** [review]

**Pull request merged**